networkit/cpp/io/FastSNAPGraphReader.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * FastSNAPGraphReader.h
+ *
+ *  Created on: 04.05.2018
+ *      Author: Alexander van der Grinten
+ */
+#ifndef FASTSNAPGRAPHREADER_H_
+#define FASTSNAPGRAPHREADER_H_
+
+#include <unordered_map>
+
+#include "../graph/Graph.h"
+#include "GraphReader.h"
+
+namespace NetworKit {
+
+/**
+ * @ingroup io
+ */
+class FastSNAPGraphReader : public NetworKit::GraphReader {
+private:
+	std::unordered_map<node, node> nodeIdMap;
+	bool directed;
+	count nodeCount;
+	bool remapNodes;
+
+public:
+	/**
+	 * Default constructor for the FastSNAPGraphReader.
+	 * NOTE: Keep in mind that many SNAP graphs do not have consecutive node ids.
+	 * Passing a nodeCount might result in a higher memory use for the output graph
+	 * but reduces runtime. Leaving this parameter on its default value, the reader
+	 * is forced to remap the nodes and thereby decreases the memory use on cost of
+	 * slighty higher runtime.
```

Comment:
This comment doesn't seem to correspond to the implementation as `nodeCount` is not used in the graph initialization in any way but only used for reserving memory for `nodeIdMap` which is not part of the final graph.

If you say this doesn't impact the performance I wonder if it really makes sense to bother the user with an additional parameter that makes usage much more complicated in many cases.

Reply:
Indeed, the comment was still for an old version. I wrote down the correct behaviour in https://github.com/kit-parco/networkit/pull/168/commits/f944ddb670fdac2161c583d90acf6a2e0414604d.

I did not remove the NodeCount parameter, because I am still hoping we can get performance increases through reserving graph space. Do you have an idea why the changes in https://github.com/rbange/networkit/commit/fb3f089f36965734fb5170286ac903163f4bd470 does not show any performance increase? I tested the reader with bigger graphs (>1 mio nodes). Is reserving space in C++ not as relevant as thought?

Reply:
The problem is here you are only reserving space for the vectors over all nodes, which is a small number of vectors. The majority of storage space is in the neighborhood and weight information - for a million nodes, you have at least a million of these vectors (assuming an unweighted, undirected graph). All these 1 million vectors need memory allocations for every resize - and there are many, as each of them will be resized several times as more and more edges are added. I assume this costs much more time than resizing the vectors over all nodes a few times (note that resizes always double the size, so this won't happen too often after a certain number of nodes have been seen). Further, sequential memory accesses are fast. Note that storing something like the 24 bytes that are needed per vector needs just 24MB for a million nodes, so resizing this vector something like 20 times shouldn't have a big performance impact (further, most of these resizes happen when the vector is still much smaller).